pipeline/transformer/transformer_calm/src/main/scala/uk/ac/wellcome/platform/transformer/calm/TransformerWorker.scala

```diff
@@ -37,60 +43,64 @@ trait TransformerWorker[In, SenderDest] {
+    transformer(sourceData, key.version) match {
+      case Left(err)     => Left(TransformerError(err.toString, sourceData, key))
+      case Right(result) => Right(result)
     }
 
-  val decodeMessage: Flow[StreamMessage, Result[(In, StoreKey)], NotUsed] =
-    Flow.fromFunction(message => decodeKey(message._2) flatMap getRecord)
-
-  val work: Flow[(In, StoreKey), Result[TransformedBaseWork], NotUsed] =
-    Flow.fromFunction {
-      case (sourceData, key) =>
-        transformer(sourceData, key.version) match {
-          case Left(err)     => Left(TransformerError(err.toString))
-          case Right(result) => Right(result)
-        }
+  private def done(work: TransformedBaseWork, key: StoreKey): Result[Unit] =
+    sender.sendT(work) toEither match {
+      case Left(err) => Left(MessageSendError(err.toString, work, key))
+      case Right(_)  => Right((): Unit)
     }
 
-  val done: Flow[TransformedBaseWork, Result[Unit], NotUsed] =
-    Flow.fromFunction(work =>
-      sender.sendT(work) toEither match {
-        case Left(err) => Left(MessageSendError(err.toString))
-        case Right(_)  => Right((): Unit)
-    })
-
-  private def decodeKey(message: NotificationMessage) =
+  private def decodeKey(message: NotificationMessage): Result[StoreKey] =
     fromJson[StoreKey](message.body).toEither match {
-      case Left(err)     => Left(DecodeKeyError(err.toString))
+      case Left(err)     => Left(DecodeKeyError(err.toString, message))
       case Right(result) => Right(result)
     }
 
-  private def getRecord(key: StoreKey) = store.get(key) match {
-    case Left(err)                     => Left(StoreReadError(err.toString))
-    case Right(Identified(key, entry)) => Right((entry, key))
-  }
+  private def getRecord(key: StoreKey): Result[(In, StoreKey)] =
+    store.get(key) match {
+      case Left(err)                     => Left(StoreReadError(err.toString, key))
+      case Right(Identified(key, entry)) => Right((entry, key))
+    }
 
   def run(): Future[Done] =
     stream.runStream(
       name,
       source => {
-        val end = source.via(Flow.fromFunction(message => message._1))
-        val processed = withSource(source)
+        source.mapAsync(2) {
+          case (message, notification) =>
+            process(notification) match {
+              case Left(err) => {
+                // We do some slightly nicer logging here to give context to the errors
+                err match {
+                  case DecodeKeyError(_, message) =>
+                    error(s"$name: DecodeKeyError from $message")
+                  case StoreReadError(_, key) =>
+                    error(s"$name: StoreReadError on $key")
+                  case TransformerError(_, sourceData, key) =>
+                    println(s"$name: TransformerError on $sourceData with $key")
+                  case MessageSendError(_, work, key) =>
+                    println(s"$name: MessageSendError on $work with $key")
```

Comment:
Can we use debug or info instead of println for these two?

Reply:
Sorry - bad debugging on my behelf - thanks!

Reply:
Using `error` though - does that make sense?